lib/compiler/src/beam_trim.erl

```diff
@@ -38,51 +39,55 @@ module({Mod,Exp,Attr,Fs0,Lc}, _Opts) ->
 
 function({function,Name,Arity,CLabel,Is0}) ->
     try
-        St = #st{safe=safe_labels(Is0, [])},
-        Is = trim(Is0, St, []),
+        St = #st{safe=safe_labels(Is0, []),fsz=0},
+        Usage = none,
+        Is = trim(Is0, Usage, St, []),
         {function,Name,Arity,CLabel,Is}
     catch
         Class:Error:Stack ->
 	    io:fwrite("Function: ~w/~w\n", [Name,Arity]),
 	    erlang:raise(Class, Error, Stack)
     end.
 
-trim([{init_yregs,{list,Kills0}}=I|Is0], St, Acc) ->
-    Kills = [{kill,Y} || Y <- Kills0],
-    try
-        %% Find out the size and layout of the stack frame.
-        %% Example of a layout:
-        %%
-        %%    [{kill,{y,0}},{dead,{y,1},{live,{y,2}},{kill,{y,3}}]
-        %%
-        %% That means that y0 and y3 are to be killed, that y1
-        %% has been killed previously, and that y2 is live.
-        {FrameSize,Layout} = frame_layout(Is0, Kills, St),
-
-        %% Calculate all recipes that are not worse in terms
-        %% of estimated execution time. The recipes are ordered
-        %% in descending order from how much they trim.
-        IsNotRecursive = is_not_recursive(Is0),
-        Recipes = trim_recipes(Layout, IsNotRecursive),
-
-        %% Try the recipes in order. A recipe may not work out because
-        %% a register that was previously killed may be
-        %% resurrected. If that happens, the next recipe, which trims
-        %% less, will be tried.
-        try_remap(Recipes, Is0, FrameSize)
-    of
-	{Is,TrimInstr} ->
-            %% One of the recipes was applied.
-	    trim(Is, St, reverse(TrimInstr)++Acc)
-    catch
-	not_possible ->
+trim([{init_yregs,_}=I|Is], none, St0, Acc) ->
+    case usage(Is, St0) of
+        none ->
+            trim(Is, none, St0, [I|Acc]);
+        {FrameSize,Us} ->
+            St = St0#st{fsz=FrameSize},
+            trim([I|Is], Us, St, Acc)
+    end;
+trim([{init_yregs,{list,Killed}}=I|Is0], [U|Us], St, Acc) ->
+    FrameSize = St#st.fsz,
+
+    %% Find out layout of the stack frame. Example of a layout:
+    %%
+    %%    [{kill,{y,0}},{dead,{y,1},{live,{y,2}},{kill,{y,3}}]
+    %%
+    %% That means that y0 and y3 are to be killed, that y1
+    %% has been killed previously, and that y2 is live.
+    Layout = frame_layout(FrameSize, Killed, U),
+
+    %% Find a trim recipe.
+    IsNotRecursive = is_not_recursive(Is0),
+    case trim_recipe(Layout, IsNotRecursive, U) of
+        none ->
             %% No recipe worked out. Use the original init_yregs/1
             %% instruction.
-	    trim(Is0, St, [I|Acc])
+	    trim(Is0, Us, St, [I|Acc]);
+        R ->
+            %% Apply the recipe.
+            {TrimInstr,Map} = expand_recipe(R, FrameSize),
+            Is = remap(Is0, Map, []),
+	    trim(Is, none, St, reverse(TrimInstr) ++ Acc)
```

Comment:
I think this would be clearer as a body-recursive function. `reverse(TrimInstr) ++ Acc` feels a bit awkward.

Reply:
Yes, make sense. Will fix.